### PR TITLE
adds OS controlled dark mode

### PR DESCRIPTION
**Pull request by jzwood**
## Changes
- removes all theme-dependent colors from default css
- adds _theme.css_ with a dark and light color palette that respects the OS theme settings

## How to Test
- run site locally
    - `stack run -- site watch`
    - navigate to site: `http://localhost:8000/`
- Chrome/Safari (macOs)
    - go to settings > appearance
        - toggle between Light and Dark mode and see that the page changes color modes
- Firefox (macOs)
    - open developer tools and toggle between sun and moon icon (see ss)
 

## Related Issue
- https://github.com/haskellfoundation/error-message-index/issues/353

### Screen Shot
Screenshot 2023-02-21 at 7 15 37 PM

### CODE REVIEW

**Comment by david-christiansen** [comment]
I'll give it a proper test soon!

As far as screenshot-based feedback goes, I have a couple questions:
 * The contrast between the light text and dark background is very high, but the same is not true for things like module names in the code sample (it also looks like the code sample's syntax highlighting is thoroughly off in both sides, but I suspect that this is unrelated to the PR). Looking at e.g. GitHub's dark theme, the contrast between the very dark background and ordinary text seems to be a bit less. How should we think about selecting the right contrast and thus text color?
 * Headers like "Examples" and "Error messages" are very low-contrast, making me think that they are not important. 
 * The pink links seem to jump out quite a bit compared to the other theme
 * Is there a distinction between visited and unvisited links? The current light theme has one, and it would be interesting to know the reasoning.

I'm not at all an expert in design, please answer my questions in ELI5 mode if possible :-) I'll also test it with the GNOME dark mode setting and Firefox - I don't have a Mac handy to test with, but that's what you're using, so this gets us a bit more coverage at least.

Thank you!

**Comment by jzwood** [comment]
Hi @david-christiansen, thanks for preliminary feedback. You brought up some great points.

I agree that the dark mode headers are too dark and the text is too bright. I made the headers brighter (by eye) and the text a bit darker (to match Github) so that feels a little better. In terms of finding a balance, it's a tad tricky b/c accessibility guidelines recommend high contrast over low but dark users tend to abhor high-contrast UI 😅.

I quite like the pink links but I'm not married to any of these colors so we can tweak as needed. That shade of pink was lifted from how [hackage](https://hackage.haskell.org/) styles links.

Re visited vs unvisited links -- my understanding is that the different [link states](https://developer.mozilla.org/en-US/docs/Learn/CSS/Styling_text/Styling_links#link_states) exist for 2 interconnected reasons.
- These distinct states are how links have worked since the beginning of the web so maintaining that behavior will make using the site more intuitive/familiar
- There are accessibility reasons for having link states be visually distinct although I admit it's unclear to me how that's applicable for specifically visited links.

_Final thoughts_: note that I've adding instructions for testing the theme with Firefox and I wanted to clarify that I have not tampered with the light theme at all. It looks like the syntax highlighting is coming from highlight.js so picking out a different theme would be a separate work issue if you wanted something prettier? 😃

**Comment by david-christiansen** [comment]
Ah, I just saw that Hackage has dark mode - perhaps the best thing to do is to just take their design and colors and run with it.

File: message-index/css/theme.css [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  a:visited {
+    color: #8B78C5;
+  }
+
+  code.hljs {
+    background-color: transparent;
+    filter: invert(1);
+  }
```
Comment (david-christiansen) [inline message-index/css/theme.css:32]: This is a nice hack, but I think it would make more sense to update `highlight.css` to include a dark theme as well, with colors chosen to match the site. I'd be willing to merge this PR without that, though, and open an issue for the future.

File: message-index/css/theme.css [hunk 2]
```diff
@@ -0,0 +1,70 @@
+body {
+  --bg-color: #FFFFFF;
+  --text-color: #333;
+  --outline-color: #DB83ED;
+  --header-color: #6D28D9;
+  --bg-nav: transparent;
+  --code-color: #FAFAFA;
```
Comment (david-christiansen) [inline message-index/css/theme.css:7]: This variable governs the _background_ color of code, but the name (by analogy to `--text-color` above) suggests that it governs the _foreground_ color. What about renaming it to `--code-bg-color`?

File: message-index/css/theme.css [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    --code-color: transparent;
+    --filename-bg: #2C2C2C;
+  }
+
+  a {
+    color: #EB82DC;
+  }
+
+  a:visited {
+    color: #8B78C5;
```
Comment (david-christiansen) [inline message-index/css/theme.css:26]: For Hackage match:
```suggestion
    color: #D5C5FF;
```

**Review by david-christiansen — CHANGES_REQUESTED** [review]
Thanks for your patience!

I'm far from an expert in CSS, so please take any suggestions with a grain of salt. Generally speaking, I really like the use of variables to structure things and make it easier to maintain over time. Variables have only been supported since 2017 in Edge, but Firefox has supported them since 2014 and most other browsers since 2016. I suppose six years is long enough to wait, Haskellers are demographically likely to have updated browsers, and the HTML is semantic enough that the whole thing should be reasonably readable with defaults.

Generally speaking, the whole CSS code became easier to read with these changes.

I left some detailed comments in the code. Thanks again!

File: message-index/css/default.css [hunk 4]
```diff
@@ -42,14 +40,10 @@ details > summary {
 
 .example {
   flex-grow: 1;
-  overflow-x: scroll !important;
+  overflow-x: auto;
```
[the remark this answers is not in the document]
Comment (jzwood) [reply message-index/css/default.css:43]: Absolutely! Using `auto` should result the scrollbar being hidden when all of the content fits into view and the scrollbar visible if there's any overflow content. When you have `overflow-x: scroll`, the scrollbar gutter shows up all the time which is undesirable b/c there's no non-hack way to style the scrollbar and on dark mode the scrollbar becomes a garish white rectangle across the screen.

Screenshot 2023-02-27 at 9 04 33 AM

when we use `auto` we instead get

Screenshot 2023-02-27 at 8 51 32 AM


Screenshot 2023-02-27 at 8 53 12 AM

[the remark this answers is not in the document]
Comment (david-christiansen) [reply message-index/css/default.css:43]: This will presumably depend on the browser and OS, but I think that your browser/OS combo is common enough that we should do this. Thanks for clearing it up!

File: message-index/css/theme.css [hunk 5]
```diff
@@ -0,0 +1,70 @@
+body {
```
[the remark this answers is not in the document]
Comment (jzwood) [reply message-index/css/theme.css:1]: Good point! This is largely a matter of taste – there's nothing stopping us from scoping on `:root` instead of `body`. I still like scoping on `body` b/c I prefer variables defined in `:root` to be immutable. I know scoping on `body` is effectively global but personally it feels a little dirty to overwrite global scope. That said, if you want me to change it I will. 🙂

File: message-index/css/default.css (hunk 4 above)
[the remark this answers is not in the document]
Comment (jzwood) [reply message-index/css/default.css:43]: My pleasure! I have to get back to work, heh, but I'll respond to the rest of your excellent comment/questions tonight!

File: message-index/css/theme.css (hunk 5 above)
[the remark this answers is not in the document]
Comment (david-christiansen) [reply message-index/css/theme.css:1]: I read a few tutorials on the subject, and they all recommend `:root` for the variables.

This project is a bit funny, in that its a web project being mostly maintained by people who are not experts in the web. Thus, a certain simplicity is important, and I think that adhering to what the docs "out there" say is a good way to make things easier for CSS n00bs (or unreformed CSS-paleolithics like myself) to maintain.

This isn't mutable, is it? My understanding is that this is just a result of CSS specificity rules and media queries, rather than mutation _per se_.

File: message-index/css/default.css (hunk 4 above)
[the remark this answers is not in the document]
Comment (david-christiansen) [reply message-index/css/default.css:43]: I really appreciate it :-) And no stress - dark mode is nice to have, but certainly not a make-or-break feature for the site. Better to keep you relaxed and having fun as a contributor!

File: message-index/css/theme.css (hunk 5 above)
[the remark this answers is not in the document]
Comment (jzwood) [reply message-index/css/theme.css:1]: Fair enough! I'll make it scoped to `:root`

File: message-index/css/theme.css [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  --bg-color: #FFFFFF;
+  --text-color: #333;
+  --outline-color: #DB83ED;
+  --header-color: #6D28D9;
+  --bg-nav: transparent;
```
[the remark this answers is not in the document]
Comment (jzwood) [reply message-index/css/theme.css:6]: Good catch! 🦅

File: message-index/css/theme.css [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+*:focus-visible {
+  outline-color: var(--outline-color);
+}
+
+body {
+  color: var(--text-color);
+  background-color: var(--bg-color);
+  text-rendering: optimizeLegibility;
+  -webkit-font-smoothing: antialiased;
+  -moz-osx-font-smoothing: grayscale;
```
[the remark this answers is not in the document]
Comment (jzwood) [reply message-index/css/theme.css:44]: Let me try and remember the motivation for the above 3 times...

[the remark this answers is not in the document]
Comment (jzwood) [reply message-index/css/theme.css:44]: ok, so those 3 lines were an attempt to make the error message more legible but that's silly -- I should just make it more legible. IMHO it's too small so in my next commit I'm going to bump it up like below.

Screenshot 2023-02-27 at 9 06 37 PM

File: message-index/css/theme.css [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+*:focus-visible {
+  outline-color: var(--outline-color);
+}
+
+body {
+  color: var(--text-color);
+  background-color: var(--bg-color);
+  text-rendering: optimizeLegibility;
```
[the remark this answers is not in the document]
Comment (jzwood) [reply message-index/css/theme.css:42]: see comments below.

File: message-index/css/theme.css [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  --bg-nav: transparent;
+  --code-color: #FAFAFA;
+  --filename-bg: #EAEAEA;
+}
+
+@media (prefers-color-scheme: dark) {
+  body {
+    --bg-color: #33333A;
+    --text-color: #C9D1D9;
+    --header-color: #8946C8;
```
[the remark this answers is not in the document]
Comment (jzwood) [reply message-index/css/theme.css:15]: We can brighten the body text for sure but I suspect that the darkened header would not pass an accessibility audit. The contrast between the header text and the background is too close:

Screenshot 2023-02-27 at 9 27 08 PM

File: message-index/css/theme.css [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    --filename-bg: #2C2C2C;
+  }
+
+  a {
+    color: #EB82DC;
+  }
+
+  a:visited {
+    color: #8B78C5;
+  }
```
[the remark this answers is not in the document]
Comment (jzwood) [reply message-index/css/theme.css:27]: The light theme uses the browser defaults for anchor tags and since I was only altering the dark mode anchor colors I thought it might be less confusing to put it directly inside the dark media query. Since the light theme might be altered I'll variablize these.

File: message-index/css/theme.css (hunk 1 above)
Comment (jzwood) [reply message-index/css/theme.css:32]: Agreed and I'm glad you appreciate the stopgap. Pulling in a real dark themed hljs will be a good next step. 👍

File: message-index/css/theme.css (hunk 2 above)
Comment (jzwood) [reply message-index/css/theme.css:7]: wise 😑

**Comment by jzwood** [comment]
Thanks for the thoughtful questions, comments, and suggestions! I think I captured the relevant feedback. Commit 1 addresses those callouts and the next one updates the header and anchor colors to match Hackage's light theme, re your suggestion. 😃 

Screenshot 2023-02-27 at 10 11 50 PM
Screenshot 2023-02-27 at 10 12 33 PM

File: message-index/css/default.css [hunk 11]
```diff
@@ -37,19 +35,15 @@ details > summary {
 }
 
 #error-message {
-  font-size: small;
+  font-size: medium;
```
Comment (jzwood) [inline message-index/css/default.css:38]: `small` is just a little too small.

File: message-index/css/theme.css (hunk 9 above)
[the remark this answers is not in the document]
Comment (david-christiansen) [reply message-index/css/theme.css]: The suggestion here was that we just do what Hackage does. It is clearly possible that I pasted the value incorrectly out of my browser dev tools :-)

File: message-index/css/theme.css (hunk 10 above)
[the remark this answers is not in the document]
Comment (david-christiansen) [reply message-index/css/theme.css]: Ah, I think it would be better if we explicitly choose colors for the light theme. I hadn't realized it didn't have values there :-) This is what we get for starting the project CSS-less and then incrementally muddling through.

File: message-index/css/theme.css (hunk 1 above)
Comment (david-christiansen) [reply message-index/css/theme.css:27]: And it looks to me like hljs just uses ordinary CSS for styling, which is great. It'll make it really easy to customize based on your skeleton here.

File: message-index/css/theme.css (hunk 2 above)
Comment (david-christiansen) [reply message-index/css/theme.css]: I'm responsible for maintaining this thing, after all :-)

File: message-index/css/default.css (hunk 11 above)
Comment (david-christiansen) [reply message-index/css/default.css:38]: Looks good!

File: message-index/css/theme.css (hunk 9 above)
[the remark this answers is not in the document]
Comment (david-christiansen) [reply message-index/css/theme.css]: And you did :-)

File: message-index/css/theme.css (hunk 1 above)
Comment (david-christiansen) [reply message-index/css/theme.css:27]: OK, #385 is opened to track this.

**Comment by david-christiansen** [comment]
Will merge as soon as CI is green. Thank you very much!

**Review by david-christiansen — APPROVED** [review]

File: message-index/css/theme.css (hunk 9 above)
[the remark this answers is not in the document]
Comment (jzwood) [reply message-index/css/theme.css]: The Hackage site faked me out too -- the header color _is_ the dark version you see on the right and they use a filter to brighten it to what you see on the left (as you noted), which is pretty confusing. It was easier to just snipe the absolute color value after it was adjusted.

**Comment by jzwood** [comment]
🥳
